Support setting Code Definition Window context on caret changes

Pilchie opened the pull request.

Pilchie commented:
Tagging @roslyn-ide and @KirillOsenkov

dpoeschl commented:
ETA failure info:
```
C:\Users\dotnet-bot\.nuget\packages\Microsoft.Net.Compilers\1.2.0-beta-20151211-01\tools\Microsoft.CSharp.Core.targets(67,5): error MSB3883: Unexpected exception:  [d:\j\workspace\dotnet_roslyn_prtest_win_dbg_eta\Open\src\Compilers\CSharp\Test\Emit\CSharpCompilerEmitTest.csproj]
C:\Users\dotnet-bot\.nuget\packages\Microsoft.Net.Compilers\1.2.0-beta-20151211-01\tools\Microsoft.CSharp.Core.targets(67,5): error : System.NullReferenceException: Object reference not set to an instance of an object. [d:\j\workspace\dotnet_roslyn_prtest_win_dbg_eta\Open\src\Compilers\CSharp\Test\Emit\CSharpCompilerEmitTest.csproj]
C:\Users\dotnet-bot\.nuget\packages\Microsoft.Net.Compilers\1.2.0-beta-20151211-01\tools\Microsoft.CSharp.Core.targets(67,5): error : at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [d:\j\workspace\dotnet_roslyn_prtest_win_dbg_eta\Open\src\Compilers\CSharp\Test\Emit\CSharpCompilerEmitTest.csproj]
```

Pilchie commented:
Hey @agocke I saw a similar failure on a PR yesterday.  Is there a crash in the new toolset we need to look for?

davkean commented:
ETA failure is: https://github.com/dotnet/roslyn/issues/7470.

retest eta please

brettfo commented on src/EditorFeatures/Core/CodeDefinitionWindowLocation.cs:
```diff
@@ -0,0 +1,20 @@
+namespace Microsoft.CodeAnalysis.Editor
+{
+    internal struct CodeDefinitionWindowLocation
+        {
```
I hope this is just extra spaces and not a tab; I'd hate for us to not be friends anymore.

KirillOsenkov replied:
Yeah I couldn't believe my eyes when I saw that. Decided my eyes are wrong and moved on.

Pilchie replied:
Mea culpa - I had this as a nested type, and moving the lines with VsVim didn't fix the indenting. :blush:

brettfo commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+            _codeDefinitionWindowService.SetContext(locations);
+        }
+
+        private async Task<ImmutableArray<CodeDefinitionWindowLocation>> GetContextFromPoint(SnapshotPoint? pointInRoslynSnapshot, TaskScheduler foregroundTaskScheduler, CancellationToken cancellationToken)
```
Maybe rename to `GetContextFromPointAfterDelay()`?  I panicked when I saw `Task.Delay()`.  Just a suggestion, certainly not blocking.

jasonmalinowski replied:
Why is pointInRoslynSnapshot nullable here? It looks like you've already filtered that out earlier, so pass .Value in the call point?

Pilchie replied:
Both changed.

brettfo commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+    [TextViewRole(PredefinedTextViewRoles.Interactive)]
+    internal class DefinitionContextConnectionListener : IWpfTextViewConnectionListener
+    {
+        private readonly Dictionary<ITextView, CancellationTokenSource> _views = new Dictionary<ITextView, CancellationTokenSource>();
```
Maybe rename?  A name like `_views` makes me think that the important thing being saved here are `ITextView`s, where it's really `CancellationToken`s.

jasonmalinowski replied:
Yes, or at least a comment. I couldn't guess at all what this is for at this point.

Pilchie replied:
renamed to `_subscribedViews` and converted to a `HashSet<ITextView>` after taking @jasonmalinowski's other suggestion of using a single task instead of one per view.

brettfo commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+        void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
```
Are there any other cases where the `CancellationTokenSource` would need to be removed from the dictionary?  I have no idea, but I don't want this to arbitrarily grow.

Pilchie replied:
There shouldn't be.  This condition is supposed to be: either the view is closed, or the projection buffer has changed so that it no longer contains any source spans that come from Roslyn documents.

balajikris commented:
Hmm, no tests?

balajikris commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+                return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+            }
+
+            var workspace = document.Project.Solution.Workspace;
```
check for cancellations inside this method ? the foreach loop seems a good candidate, there could be other places as well.

Pilchie replied:
I actually think this is okay - I mostly wanted to have the `CancellationToken` for the delay at the beginning and the potentially expensive operations like finding the symbol, mapping it between workspaces, etc.  Nothing in those loops is particularly expensive.

balajikris commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+        private async Task UpdateDefinitionContext(ITextView textView, CaretPosition caretPosition)
+        {
+            // Cancel any pending update for this view
+            _views[textView].Cancel();
```
you need to check for cancellations inside `GetContextFromPoint` for this to be effective?

balajikris commented on src/VisualStudio/Core/Def/Implementation/VisualStudioCodeDefinitionWindowService.cs:
```diff
@@ -0,0 +1,99 @@
+
+using Microsoft.CodeAnalysis.Editor;
+using System;
+using System.Linq;
```
Place System on top?

Pilchie replied:
Doh!  Thanks!

balajikris commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+
+            var workspace = document.Project.Solution.Workspace;
+            if (!document.SupportsSemanticModel)
+            {
```
What documents are these? .ts? are there others?

balajikris replied:
Related Q: Documents in Misc files workspace, do they take the else branch, are they expected to work?

Pilchie replied:
Yes, I expect these are .ts files.  I copied this pattern from Peek, but I should definitely check the behavior with a TS file open.

For the related question, yes, they take the else branch, and they work because the symbol mapping service will figure out where they came from originally.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+
+        void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
```
If the buffer is being disconnected from the buffer graph we don't care? Why not?

balajikris commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+        {
+            if (!_views.ContainsKey(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _views.Add(textView, new CancellationTokenSource());
```
I think this `CancellationTokenSource` is never used, `UpdateDefinitionContext` creates a new one anyway?

Pilchie replied:
I've removed it now, but it was there so that I didn't have to check for null in `UpdateDefinitionContext` for that very first time.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+
+            // After a delay in case the caret moves again, find the symbol under the caret and update the context
+            var cancellationTokenSource = new CancellationTokenSource();
+            _views[textView] = cancellationTokenSource;
```
Is there some synchronization on _views I'm missing here? What if two threads do this?

jasonmalinowski replied:
Ah, the assumption is that this is foreground affinitized? Given it's saving you just the single marshal from the background thread to the foreground thread to call SetContext, what about just doing an old fashioned .ContinueWith instead?

Pilchie replied:
`_views` is (now) only accessed in `SubjectBuffersConnected` and `SubjectBuffersDisconnected`, so it should be safe.

However, I suggest that you look more closely at the threading of `GetContextFromPoint()`, since it does something I don't really like.

It starts out on a threadpool thread, but eventually needs to call `ISymbolNavigationService.WouldNavigateToSymbol()` which has to happen on the UI thread.  That is the one place where it uses `foregroundTaskScheduler`.

A close examination of my use of `ConfigureAwait()` and related threading operations would be appreciated.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+            var foregroundTaskScheduler = TaskScheduler.FromCurrentSynchronizationContext();
+
+            var locations = await Task.Run(
+                async () => await GetContextFromPoint(pointInRoslynSnapshot, foregroundTaskScheduler, cancellationToken).ConfigureAwait(false),
```
Can't you just TaskRun(() => GetContextFromPoint())? and skip the extra wrapping of async machinery?

Pilchie replied:
Yep. Done

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+        {
+            // TODO: Does this allocate too many tasks - should we switch to a queue like the classifier uses?
+            await Task.Delay(TimeSpan.FromMilliseconds(250), cancellationToken).ConfigureAwait(false);
+            var document = pointInRoslynSnapshot?.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
```
We used to have some optimizations where it was desirable to call this and capture it earlier rather than doing it asynchronously, since we'd have some stuff cached that otherwise might have to be reconstructed. Is that still the case?

Pilchie replied:
I was unaware of said optimizations - I did this to avoid the `WithChanges` part of the call needing to fork the workspace early, but I'm happy to change if you think it's better.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+            }
+            else
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```
Do we have our own GoToDefinition service here that already does this entire branch? Seems odd we'd switch like this.

Pilchie replied:
I patterned this off of Peek, so if we do, either we didn't when you wrote Peek, or Peek wasn't updated to use it :stuck_out_tongue:.  Will take a look.

Pilchie replied:
tl;dr is that we don't.

Long answer is that we implement it, but it doesn't do any of the good stuff(metadata as source, symbol mapping, XAML mapping, etc) for C#/VB in the `FindDefinitionsAsync()` method, only in `TryGoToDefinition()`.

Pilchie replied:
The only two callers of FindDefinitionsAsync() are Peek and this though, so it seems like we should be able to try an unify them into that method a bit.  I'll look more at that later.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+                var symbol = SymbolFinder.FindSymbolAtPosition(
+                    semanticModel,
+                    pointInRoslynSnapshot.Value.Position,
+                    workspace, bindLiteralsToUnderlyingType: true,
```
Split line on comma.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+                }
+
+                symbol = symbol.GetOriginalUnreducedDefinition();
+                var project = document.Project;
```
It seems like this is leaving a gotcha where "project" and "document" may be from entirely different workspaces.

Pilchie replied:
Extracted remainder out so that it doesn't have access to document.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+                        foreach (var declaration in sourceLocations)
+                        {
+                            var declarationLocation = declaration.GetLineSpan();
+                            results.Add(new CodeDefinitionWindowLocation(symbol.ToDisplayString(), declarationLocation.Path, declarationLocation.Span.Start.Line, declarationLocation.Span.Start.Character));
```
Should your CodeDefinitionWindowLocation type just use something else for line/column? It seems every place you construct one you have to deconstruct some structure.

Pilchie replied:
A couple of constructor overloads helped it a bunch.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/VisualStudioCodeDefinitionWindowService.cs:
```diff
@@ -0,0 +1,99 @@
+
+        public void SetContext(ImmutableArray<CodeDefinitionWindowLocation> locations)
+        {
+            // If the new context has no location, then just don't update, instead of showing the
```
Why is the policy being decided here? Shouldn't that be in the feature code?

Pilchie replied:
What's your proposal for the contract?  Empty locations means reset, and the caller chooses not to call SetContext if they don't want it to change?

jasonmalinowski replied:
Oh that's the intent? How about two methods, SetContext (must be non-empty list) and Clear?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/VisualStudioCodeDefinitionWindowService.cs:
```diff
@@ -0,0 +1,99 @@
+
+            int IVsCodeDefViewContext.GetCol(uint iItem, out uint piCol)
+            {
+                var index = (int)iItem;
```
Extract the "get item and throw if wrong" pattern?

Pilchie replied:
Done

jasonmalinowski commented:
I'm surprised that there's potentially a task running per view instead of a single task running globally. It seems if I switch to another view I wouldn't care anymore about the first task running. Even worse, if that task gets delayed for some reason, then a navigation in the view I navigated to might get blown away by that delayed task. It'd probably also help performance too.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+            if (!_views.ContainsKey(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _views.Add(textView, new CancellationTokenSource());
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
```
Note this only covers _explicit_ repositions.

> The event is not raised if the caret position was changed as a consequence of tracking normal text edits. The normal behavior of the caret is to move after the typed character.

I assume if I type the name of a type and then pause, it should show the definition of that. To that end, you would need to watch more events.

Pilchie replied:
I agree that we want that - do you know of a place where we do something like this?

jasonmalinowski replied:
The only place we do this sort of stuff is in taggers, and we have our event source stuff to do that all for us. Just also subscribe to PostChanged on the view's buffer as well when you subscribe to CaretMoved? Should be fairly sane I think.

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+                async () => await GetContextFromPoint(pointInRoslynSnapshot, foregroundTaskScheduler, cancellationToken).ConfigureAwait(false),
+                cancellationToken).ConfigureAwait(true);
+
+            _codeDefinitionWindowService.SetContext(locations);
```
Do you need to recheck the cancellation token here, since it's possible that this operation was canceled while we redispatched to the UI thread?

Pilchie replied:
Done.

Pilchie commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,211 @@
+                }
+
+                var solution = project.Solution;
+                var sourceDefinition = await SymbolFinder.FindSourceDefinitionAsync(mappingResult.Symbol, mappingResult.Project.Solution, cancellationToken).ConfigureAwait(false);
```
I was seeing a `NullReferenceException` here while playing with this tonight.  Must be a case where `mappingResult` is non-null but either `Symbol` or `Project` is.

jasonmalinowski replied:
Does MapSymbolAsync not just return a mapping result with the inputs if it can't do it's magic? Should it? The pattern seems really funky otherwise...

Pilchie replied:
I haven't had a chance to debug yet.  Note that I copied this code from Peek, but couldn't repro the same null ref there in a few minutes of trying, so maybe something else was up.

Pilchie replied:
Okay, the issue here is that mapping fails entirely when the *source* workspace is a `MetadataAsSourceWorkspace` due to #6791.  Other callers of `MapSymbolAsync` guard against null returns, so I'll do that too.

Pilchie commented:
I believe I've addressed all PR feedback (except the lack of tests :blush:), either by resolving it or by asking for follow-up.  Would appreciate another round of :eyes: on this, especially as it relates to threading @jasonmalinowski

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,215 @@
+
+        private async Task UpdateDefinitionContext(ITextView textView, CaretPosition caretPosition)
+        {
+            // Cancel any pending update for this view
```
"any view"

jasonmalinowski commented on src/EditorFeatures/Core/DefinitionContextTracker.cs:
```diff
@@ -0,0 +1,215 @@
+            {
+                if (_subscribedViews.Contains(textView))
+                {
+                    _subscribedViews.Remove(textView);
```
What happens if there's an active task computing the code definition for that view?

jasonmalinowski commented:
:+1:

DustinCampbell commented:
@Pilchie: We're coming up on the anniversary here. What's the status of this PR?

jaredpar commented:
We apologize, but we are closing this PR due to code drift. We regret letting this PR go so long without attention, but at this point the code base has changed too much for us to revisit older PRs. Going forward, we will manage PRs better under an SLA currently under draft in issue #26266 – we encourage you to add comments to that draft as you see fit.   Although that SLA is still in draft form, we nevertheless want to move forward with the identification of older PRs to give contributors as much early notice as possible to review and update them.

If you are interested in pursuing this PR, please reset it against the head of master and we will reopen it. Thank you!

jasonmalinowski commented:
We discussed this in the design meeting and couldn't come up with any reason not to take this (obviously with a bunch of updates to modernize the code.)
